source/functions/handle-errors.txt

```diff
@@ -0,0 +1,422 @@
+.. _functions-handle-errors:
+
+==========================
+Handle Errors in Functions
+==========================
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+This page explains how to work with errors in Atlas Functions.
+
+.. _functions-basic-error-handling:
+
+Basic Error Handling
+--------------------
+
+You can handle Function errors using standard JavaScript error handling techniques
+like :mdn:`try...catch statements <Web/JavaScript/Reference/Statements/try...catch>`.
+
+.. literalinclude:: /examples/generated/functions/willThrowAndHandleError.snippet.handle-error-try-catch.js
+   :language: js
+
+.. _functions-view-logs:
+
+View Logs
+---------
+
+You can view records of all Function executions including which an error prevented
+successful execution in :ref:`App Service Logs <logs>`.
+
+Depending on how a Function is invoked, it shows up differently in the logs.
+For example, logs for Functions called by Atlas Triggers have a different format
+from logs for Functions called directly from a Realm client SDK.
+For more information, refer to the :ref:`Log entry type documentation <log-entry>`.
+
+.. _functions-retry:
+
+Retry Functions
+---------------
+
+Atlas Functions do not have built-in retry behavior.
+You can add custom retry behavior. For example, you might want to add
+retry behavior if the third-party service that your Function calls has
+intermittent connectivity, and you want the Function to re-execute
+even if the third-party service is temporarily down.
+
+This section describes the following that you can add retry behavior to your Functions:
+
+- :ref:`Recursively Call Functions in Error Handling Blocks <functions-recursive-call>`
+- :ref:`Use Database Triggers to Retry Functions <functions-database-trigger-retry>`
+
+.. _functions-recursive-call:
+
+Recursively Call Function in Error Handling Blocks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One way you can retry functions is to handle operations that might fail by
+calling a Function :wikipedia:`recursively <Recursion_(computer_science)>`.
+
+On a high-level, this process includes the following components:
+
+- Execute operations that you want to retry in a ``try`` statement
+  and have the Function call itself in a ``catch`` statement.
+- So that the execution doesn't happen indefinitely, set a maximum number of retries.
+  Every time that the Function fails and enters the ``catch`` statement,
+  increment a count of the current number of retries.
+  Stop the recursive execution when the Function's current number of retries
+  reaches the max number of retries.
+- You may also want to `throttle <https://docs.microsoft.com/en-us/azure/architecture/patterns/throttling>`__
```

Comment:
Similar to my note above, I wonder if throttling is really a concept we need to explain in our docs, or if this is an unnecessary link off of our docs site. I think we should probably avoid linking out for general computer science concepts, but it can be useful to link out for specific framework or architecture-specific info that readers may not be familiar with. Throttling feels general to me.